## Quiz Examples/Questions (week 1)

In [1]:
import numpy as np
import matplotlib.pyplot as plt

from scipy.stats import binom

In [2]:
# Calculate likelihood based on binomial distribuition
def binomial_likelihood(k, n, p): 
    likelihood = []
    for probability in p: 
        l = binom.pmf(k, n, probability)
        likelihood.append(l)
    return likelihood

In [8]:
# Calculate posterior
def calculate_posterior(likelihood, priors): 
    posterior = []
    numerator = []
    denominator = 0
    sum_posterior = 0

    for idx, l in enumerate(likelihood): 
        n = l*priors[idx]
        numerator.append(n)
        denominator += n

    for n in numerator: 
        post = n/denominator
        posterior.append(post)
        sum_posterior += post
    return posterior

***
> You are told that a coin has one of the following, with the probability of heads under that event noted next to it in parentheses:    
> - a strong tails bias (p = 0.2)   
> - a weak tails bias (p = 0.4)   
> - no bias (p = 0.5)   
> - a weak heads bias (p = 0.6)   
> - a strong heads bias (p = 0.8)   
> You assign a prior probability of 1/2 that the coin is fair and distribute the remaining 1/2 prior probability equally over the other four possible scenarios. You flip the coin three times and it comes up heads all three times. What is the posterior probability that the coin is biased towards heads?

In [18]:
priors = [0.125, 0.125, 0.5, 0.125, 0.125] 
p = [0.2, 0.4, 0.5, 0.6, 0.8]
likelihood = binomial_likelihood(3, 3, p)
likelihood

[0.008000000000000004,
 0.06400000000000002,
 0.12500000000000003,
 0.21599999999999997,
 0.5120000000000001]

In [20]:
posterior = calculate_posterior(likelihood, priors)
posterior

[0.006153846153846156,
 0.04923076923076923,
 0.38461538461538464,
 0.1661538461538461,
 0.39384615384615385]

Strong heads bias + Weak heads bias = Bias towards head    
0.3938 + 0.1661 $\approx$ 0.56


***
> Julia is having an outdoor wedding ceremony tomorrow. In recent years, it has rained on average 50 days per year. Unfortunately, the meteorologist has predicted rain for her wedding day. When it rains, the meteorologist will have correctly predicted it 80 percent of the time. When it does not rain, the meteorologist will have incorrectly predicted rain 30 percent of the time. Given this information, what is the probability that it rains on Julia's wedding day?

Need to include figure (decision)

*** 
> A New York City cab was involved in a hit-and-run accident last night. Five witnesses reported the incident, four of whom said that the cab was green and one of whom said that the cab was yellow. Assume each witness correctly identifies the color of a cab with probability 2/3. It is known that 85% of registered cabs in New York City are yellow and 15% are green. Based on this information, what is the probability that the cab was green?


Understanding the problem:    
    
In this example, 4 people out of 5 saw the cab as green so we have n=5 and k=4.    
Our prior beliefs of the cab being green or yellow are respectively: 0.15 and 0.85   
The probability associated with the witness identifying the car color correctly is 2/3.   

This can be translated to:    
n=5, k=4, p=2/3 and prior=0.15 

In [12]:
likelihood = binomial_likelihood(4,5, [2./3., 1./3.])
likelihood

[0.32921810699588466, 0.041152263374485576]

In [13]:
priors = [0.15, 0.85]
posterior = calculate_posterior(likelihood, priors)
posterior

[0.5853658536585367, 0.41463414634146334]

Posterior $\approx$ 0.58

***
> You go to Las Vegas and sit down at a slot machine. You are told by a highly reliable source that, for each spin, the probability of hitting the jackpot is either 1 in 1,000 or 1 in 1,000,000, but you have no prior information to tell you which of the two it is. You play ten times, but do not win the jackpot. What is the posterior probability that the true odds of hitting the jackpot are 1 in 1,000?

**Hypotheses:** $H_1$: 0.001  $\qquad$ $H_2$: 0.000001     
**Prior:** $P(H_1)$=0.5 $\qquad$  $P(H_2)$=0.5    
Since we don't have any reason to believe one hypotheses is more likely than the other, we can place a 0.5 probability on each one.    

In [10]:
priors = [0.5, 0.5]
likelihood = binomial_likelihood(0,10, [0.001, 0.000001])
likelihood

[0.9900448802097482, 0.9999900000449998]

In [11]:
posterior = calculate_posterior(likelihood, priors)
posterior

[0.49750126996920313, 0.5024987300307969]

***
> You decide to conduct a statistical analysis of a lottery to determine how many possible lottery combinations there were. If there are N possible lottery combinations, each person has a 1/N chance of winning. Suppose that 413,271,201 people played the lottery and three people won. You are told that the number of lottery combinations is a multiple of 100 million and less than 1 billion, but have no other prior information to go on. What is the posterior probability that there were fewer than 600 million lottery combinations?

In [22]:
likelihood = binomial_likelihood(3, 413271201, [1./100000000, 
                                                1./200000000, 
                                                1./300000000, 
                                                1./400000000, 
                                                1./500000000,
                                                1./600000000,
                                                1./700000000,
                                                1./800000000,
                                                1./900000000])
likelihood

[0.18868617897428305,
 0.18623320340460062,
 0.10988008982130686,
 0.06541398851936421,
 0.04117952113076139,
 0.027350463308820804,
 0.019004567284508906,
 0.013706683822420287,
 0.010195365504294856]

In [23]:
posterior = calculate_posterior(likelihood, [0.111, 0.111, 0.111, 0.111, 0.111, 0.111, 0.111, 0.111, 0.111])
posterior

[0.28517518530780456,
 0.2814678244060024,
 0.16606979454865214,
 0.09886493223372123,
 0.062237613974641434,
 0.041336750178243516,
 0.028722988755806724,
 0.02071591104479858,
 0.015408999550329311]

Being less than 600 million: 

In [26]:
result = 0
for idx, r in enumerate(posterior): 
    if idx == 5: 
        break
    result += r
result    

0.8938153504708217

Using a more intuitive approach:   
Likelihood can be expressed by P(3 winners out of 413271201 | Fewer than 600 million lottery combinations)   
    
binom.pmf(3,413271201,1/100000000) + binom.pmf(3,413271201,1/200000000) + binom.pmf(3,413271201,1/300000000) + binom.pmf(3,413271201,1/400000000) + binom.pmf(3,413271201,1/500000000) $\approx$ 0.59
     
Since we don't have any other information about our prior, we can assume that the chances of any of the combinations are equally possible to occur: 1/9    
We still have to compute the denominator for the posterior, which will be the sum of all possible lottery combinations at the binomial distribuition times the prior. 


*** 
> You are testing dice for a casino to make sure that sixes do not come up more frequently than expected. Because you do not want to manually roll dice all day, you design a machine to roll a die repeatedly and record the number of sixes that come face up. In order to do a Bayesian analysis to test the hypothesis that p = 1/6 versus p = .175 , you set the machine to roll the die 6000 times. When you come back at the end of the day, you discover to your horror that the machine was unable to count higher than 999. The machine says that 999 sixes occurred. Given a prior probability of 0.8 placed on the hypothesis p = 1/6 , what is the posterior probability that the die is fair, given the censored data? 

Hint - to find the probability that at least x sixes occurred in N trials with proportion p (which is the likelihood in this problem), use the R command : 1-pbinom(x-1,N,p)

In [35]:
#H_1 = 1./6. prior = 0.8 
#H_2 = .175 prior = 0.2 
#n = 6000 k = 999 

priors = [0.8, 0.2]
likelihood = np.array(binomial_likelihood(998,6000, [1./6., 0.175]))
likelihood = 1 - likelihood
 
posterior = calculate_posterior(likelihood, priors)
posterior

[0.7982270289973356, 0.20177297100266445]